# 关于线性回归算法的更多思考

In [1]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn import datasets
boston = datasets.load_boston()
X = boston.data
y = boston.target
X = X[y < 50.0]
y = y[y < 50.0]
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 666)

In [2]:
from sklearn.linear_model import LinearRegression

In [3]:
line_regression = LinearRegression() # 多元线性回归

In [4]:
line_regression.fit(X, y) # 多元线性回归，直接传入整体的X和y

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [5]:
line_regression.coef_

array([-1.06715912e-01,  3.53133180e-02, -4.38830943e-02,  4.52209315e-01,
       -1.23981083e+01,  3.75945346e+00, -2.36790549e-02, -1.21096549e+00,
        2.51301879e-01, -1.37774382e-02, -8.38180086e-01,  7.85316354e-03,
       -3.50107918e-01])

In [6]:
line_regression.intercept_

32.229841669933634

In [7]:
np.argsort(line_regression.coef_) # 排名后按照索引显示出来

array([ 4,  7, 10, 12,  0,  2,  6,  9, 11,  1,  8,  3,  5], dtype=int64)

In [8]:
boston.feature_names

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

In [9]:
boston.feature_names[np.argsort(line_regression.coef_)] # 线性回归法具有对数据的可解释性,依次代表影响房价的因素，从轻到重排序

array(['NOX', 'DIS', 'PTRATIO', 'LSTAT', 'CRIM', 'INDUS', 'AGE', 'TAX',
       'B', 'ZN', 'RAD', 'CHAS', 'RM'], dtype='<U7')

In [10]:
print(boston.DESCR)

.. _boston_dataset:

Boston house prices dataset
---------------------------

**Data Set Characteristics:**  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pu

## 线性回归算法总结
+ 典型的参数学习
  > 对比kNN：为非参数学习
+ 只能解决回归问题
  + 但是很多分类方法中，线性回归是基础(比如逻辑回归)
  + 对比kNN:既可以解决分类问题，又可以解决回归问题
+ 对数据有假设：线性
  + 对比kNN：对数据没有假设
+ 优点：对数据有强解释性
+ 缺点：时间复杂度高，基本是O(n^3)，即使优化也就才能到O(n^2.4)
  > 为了解决这个时间复杂度过高的问题，有了下章的"梯度下降法"